### PR TITLE
Add support for multi line party finder notes

### PR DESCRIPTION
**Pull request by lineargraph**

### CODE REVIEW

**Comment by cow-mc** [comment]
It's been a few months now, but I can't re-create multi-line party notes. The input for it is an anvil GUI, so it's quite limited by the number of allowed characters.
Filling the whole notes with `#`s or `_`s will move the notes to a new line. But I can't create more than just a single line of party notes.

## single line party notes

```java
"§7§7Note: §fSome text"
```

## long party notes
```java
"§7§7Note:", "§7§f##############################"
```
The `"§7§7Note:"` line seems to lose its trailing space, however, it's still just a single line of notes

**Comment by lineargraph** [comment]
Could reproduce the multi line note like so: 
```
{
    id: "minecraft:skull",
    Count: 1b,
    tag: {
        SkullOwner: {
            Id: "d3cb85e2-3075-48a1-b213-a9bfb62360c1",
            hypixelPopulated: 1b,
            Properties: {
                textures: [{
                    Signature: "QJrOGkdY3nsJU8O+6ImyTBSPt6xD/jRpzugZH2A9iM841ruekrEnXvNP4kUM7DIQ+A4IUKr8Ub7akvQme/H+CNYpl5L/4Co97xrXWUUliEEcPIrng0x2zJuO6J4LreUOCaw4i7SkloQs6TkdyeegI2jwbai42IC9YSmFX4dLupw8sxEn9PShg3v3xU0U2LAvQQYoyaOyh8WlyEDhGiFWBwSj5hyY/RbFWQDtcu2wMxZdMO0C31IfF8toMl7qKyU7VM7rK3MlPdJoDpDWeHQRr/gJBo8PEmmHiHDrUfPG3PC56EsZbO6W575jEd99fVRyKKa59aVRHEZxdrwy4R4MwAoCkkaHny0VeaGI+qoCetPn9wOpKAOqUbtptT8RCcVzVN6ziaMMAXpuxsfUY9GN1p+gS3Qj+U5R6nEWdF0CMgrziAOcSy3x/auXHyohpv+buc0IA9fxRDqWCs1wfe47UxhPhnHFvRhSU4VrMGY2nl/Aj9ejzMns9CeM9mzE51VGtEwUNYvpdLgIAsHBJUAHdCXIyI8mFuQ4pHmLW2yxXZKIbloCR5babdxCW8Lb7QjUGdF0vRq/sDy9xDTkNNeYOL78q+DCSlssj46m6mr8VJTVELKpTYULl9kd8FpOoCUpiLwmkxtI6PTPyF1opTJHpYuEkLRfZoZxLORVqQ4nggo=",
                    Value: "ewogICJ0aW1lc3RhbXAiIDogMTcwNjQ2Nzk4MDcwNCwKICAicHJvZmlsZUlkIiA6ICJkM2NiODVlMjMwNzU0OGExYjIxM2E5YmZiNjIzNjBjMSIsCiAgInByb2ZpbGVOYW1lIiA6ICJscmc4OSIsCiAgInNpZ25hdHVyZVJlcXVpcmVkIiA6IHRydWUsCiAgInRleHR1cmVzIiA6IHsKICAgICJTS0lOIiA6IHsKICAgICAgInVybCIgOiAiaHR0cDovL3RleHR1cmVzLm1pbmVjcmFmdC5uZXQvdGV4dHVyZS81ODQxYTE2YTViZDRhNjQ2Y2VkYjRiNTQzNzcyMzIyNmM3Y2Y5Zjg2NjllNTU4NzczZmFlMGE5NDUyYzk0ZDkwIiwKICAgICAgIm1ldGFkYXRhIiA6IHsKICAgICAgICAibW9kZWwiIDogInNsaW0iCiAgICAgIH0KICAgIH0KICB9Cn0="
                }]
            },
            Name: "hypixel_skull"
        },
        display: {
            Lore: ["§7Dungeon: §bThe Catacombs", "§7Floor: §bEntrance", "§7§7Note: §faaaaaaaaaaaaaaaaaa", "§faaaaaaaaaa", "§7Dungeon Level Required: §b32", "", "§f§7Members: ", " §blrg89§f: §eArcher§b (§e27§b)", "§8 Empty", "§8 Empty", "§8 Empty", "§8 Empty", "", "§cYou are in this party!"],
            Name: "§6§blrg89's Party"
        }
    },
    Damage: 3s
}
```

But that seems to be slightly different to what it was when i wrote this updated parser as well.

**Comment by cow-mc** [comment]
I did some more testing:
```java
"§7§7Note: §fa a a a a a a a a a a a a a a", // = one line
"§7§7Note: §faaa aaa aaa aaa aaa aaa aaa", "§faa", // = last 2 chars in new line
"§7§7Note: §faaaaaaaaaa aaaaaaaaaa", "§faaaaaaaa", // = 2 lines
"§7§7Note: §faaaaaaaaaaaaaaaaaa", "§faaaaaaaaaaa", // = 2 lines
"§7§7Note:", "§7§faaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", // = all in 2nd line, no space after `Note:`, but with `§7` in the beginning of the next line
```

Minecraft 1.8.9's anvils have a limit of 30 characters. MC 1.17 and newer allow up to 50 characters, however, Hypixel limits the party notes to just 32 chars.
Maybe the 32 char limit got introduced _after_ your initial testing? So it used to allow up to 3 lines of notes when using mc 1.17 or newer, but now it's just either
1. all in one line
2. split into two lines
2. all text in a new line

File: src/main/java/de/cowtipper/cowlection/listener/skyblock/DungeonsListener.java [hunk 1]
```diff
@@ -360,7 +362,13 @@ private void renderPartyStatus(ItemStack item, int x, int y) {
...
                     --partySize;
                 } else if (toolTipLineWithoutFormatting.startsWith("Note: ")) {
-                    String partyNote = toolTipLineWithoutFormatting.toLowerCase();
+                    String partyNote = toolTipLineWithoutFormatting.toLowerCase().trim();
+                    while (i + 1 < itemTooltip.size()) {
+                        String next = EnumChatFormatting.getTextWithoutFormattingCodes(itemTooltip.get(i + 1)).trim().toLowerCase();
+                        if (next.isEmpty()) break;
+                        partyNote += " " + next;
+                        i++;
+                    }
```
Comment (cow-mc) [inline src/main/java/de/cowtipper/cowlection/listener/skyblock/DungeonsListener.java]: Something like this should work:

```suggestion
                } else if (toolTipLineWithoutFormatting.startsWith("Note:")) {
                    String partyNote = toolTipLineWithoutFormatting.substring(5 /* remove "Note:" */).trim().toLowerCase();

                    if (i + 1 < itemTooltip.size()) {
                        String nextToolTipLine = itemTooltip.get(i + 1);
                        if (!nextToolTipLine.isEmpty() && (nextToolTipLine.startsWith("§7§f") || nextToolTipLine.startsWith("§f"))) {
                            partyNote += " " + EnumChatFormatting.getTextWithoutFormattingCodes(nextToolTipLine).trim().toLowerCase();
                            i++;
                        }
                    }
```

**Comment by cow-mc** [comment]
1. The only lore line that starts with `§7§f` is when some party notes don't contain a space, so the notes are all in the 2nd line of party notes
2. The only lore line that starts with `§f` is when some party notes span 2 lines
    - the only other party lore line that starts with `§f` is the Members line (`"§f§7Members: "`), but there is always an empty line immediately before it, so it's never the line immediately after `Notes:`

**Pull request merged**